skglm/estimators.py

```diff
@@ -26,6 +27,123 @@
 )
 from skglm.utils import compiled_clone
 from skglm.solvers import cd_solver_path, bcd_solver_path
+from skglm.solvers.cd_solver import cd_solver
+
+
+def _glm_fit(X, y, model, datafit, penalty):
+
+    is_classif = False
+
+    for base in model.__class__.__bases__:
+        if base.__name__ in ["ClassifierMixin", "LinearClassifierMixin"]:
+            is_classif = True
+
+    if is_classif:
+        check_classification_targets(y)
+        enc = LabelEncoder()
+        y = enc.fit_transform(y)
+        model.classes_ = enc.classes_
+        n_classes_ = len(model.classes_)
+        sparse_format = "csc" if isinstance(datafit, Logistic) else "csr"
+        is_sparse = issparse(X)
+        if n_classes_ <= 2:
+            y = 2 * y - 1
+        X = check_array(
+            X,
+            accept_sparse=sparse_format,
+            dtype=np.float64)
+
+        y = check_array(y, ensure_2d=False, dtype=X.dtype.type)
+        check_consistent_length(X, y)
+    else:
+        check_X_params = dict(
+            dtype=[np.float64, np.float32], order='F',
+            accept_sparse='csc', copy=model.fit_intercept)
+        check_y_params = dict(ensure_2d=False, order='F')
+
+        X, y = model._validate_data(
+            X, y, validate_separately=(check_X_params, check_y_params))
+        X = check_array(X, 'csc', dtype=[np.float64, np.float32],
+                        order='F', copy=False, accept_large_sparse=False)
+        y = check_array(y, 'csc', dtype=X.dtype.type, order='F', copy=False,
+                        ensure_2d=False)
+    if y.ndim == 2 and y.shape[1] == 1:
+        warnings.warn("DataConversionWarning('A column-vector y"
+                      " was passed when a 1d array was expected")
+        y = y[:, 0]
+
+    if not hasattr(model, "n_features_in_"):
+        model.n_features_in_ = X.shape[1]
+
+    n_samples = X.shape[0]
+    if n_samples != y.shape[0]:
+        raise ValueError("X and y have inconsistent dimensions (%d != %d)"
+                         % (n_samples, y.shape[0]))
+
+    if not model.warm_start or not hasattr(model, "coef_"):
+        model.coef_ = None
+
+    penalty = compiled_clone(penalty)
+    datafit = compiled_clone(datafit, to_float32=X.dtype == np.float32)
+    if issparse(X):
+        datafit.initialize_sparse(X.data, X.indptr, X.indices, y)
+    else:
+        datafit.initialize(X, y)
+
+    if isinstance(datafit, QuadraticSVC):
```

Comment:
these cases are special cases (rep WeighedLasso). I think they should be treated inside the model rather than a function with if/else statements.

Reply:
I am not sure that we will be able to put in the datafit for the SVM.
The problem is that the SVM is solved in the dual, if we really want to take care of in this PR, I think the cleanest way would be to add a `solved_in_dual` parameter

Reply:
You can't treat that in the model because this function is called by GeneralizedLinearEstimator